### PR TITLE
Closes #1226 Adding Playbook Cache Feature

### LINKED ISSUE #1226: Adding Playbook Cache Feature
From @0x0elliot 's GSoC plan:

After every scan which uses more than one Analyzer and Connectors, We would
present the user with a pop-up that would say “Do you want to save this scan as a
Playbook?”

If the user agrees to this, The configuration would be sent to the backend where it would
be saved “lazily” in the database. The database model for each entry would somewhat
look like this:

```
class CachedPlaybook(models.Model):
name = models.Charfield(max_length=225)

class CachedPlaybookEntries(models.Model):
playbook = models.ForeignKey(to=CachedPlaybook, on_delete=models.CASCADE)
type = models.CharField(max_length=50) # analyzer/connector
name = models.CharField(max_length=225)
active = models.BooleanField()
# ^ if False, then it means that analyzer/connector
# has been altered
```

This would be in the `models.py`.

The result of the cached playbooks would be appended to the JSON returned by
`/api/playbook/playbook_configuration`.

Some further technical support would be required to be added internally for the other
endpoints so that we would be looking for the playbook information in both the playbook
JSON file and the cached database entries.

If the playbook entry has an analyzer/connector then we would be marking the cached
playbook entry as False so that the entry is ignored.

### PR DESCRIPTION
**Pull request by 0x0elliot**
# Description

Adding playbook caching after each scan.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] The tests gave 0 errors.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](./Contribute.md)).
- [x] If the analyzer is free, Please add it in the `FREE_TO_USE_ANALYZERS` playbook in `playbook_config.json`

# Real World Example
Screenshot 2022-11-16 at 4 59 24 AM

# To Dos:

- [x] Finish serializers and models.
- [x] Make sure that every new playbook that is created is unique 
- [x] Make the cache endpoint work
- [x] Append the results of these changes to PlaybookConfigSerializer's results.
- [x] And the one in the JSON file is given more priority if the same named one is ever added back again.
- [x] Finish frontend changes.
- [x] Finish test cases

### CODE REVIEW

**Comment by 0x0elliot** [comment]
by now, the APIs work. will now work on the playbooks serializer part.

**Comment by 0x0elliot** [comment]
backend should be solid.

**Comment by 0x0elliot** [comment]
frontend has some work left. will get to it by this week.

**Comment by 0x0elliot** [comment]
don't mind the linting errors. i am working on the frontend currently. that's just the ready skeleton.

**Comment by 0x0elliot** [comment]
And, we are ready for a review!

File: api_app/playbooks_manager/models.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    name = models.CharField(max_length=225, primary_key=True)
+    # Required fields
+    description = models.CharField(max_length=225, default="", blank=True)
+    analyzers = models.JSONField(default=dict)
+    connectors = models.JSONField(default=dict)
+
+    # Optional Fields
+    supports = models.JSONField(default=list)
+    disabled = models.BooleanField(default=False)
+    job = models.ForeignKey(Job, on_delete=models.CASCADE, related_name="job")
```
[the remark this answers is not in the document]
Comment (0x0elliot) [reply api_app/playbooks_manager/models.py:19]: hmm. weird. i added the suggestions it made and yet the problems are persisting. i really am not sure what it is referring to anymore.

**Comment by 0x0elliot** [comment]
will fix the test cases by today.

File: intel_owl/settings/django.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
     "certego_saas.ext.middlewares.StatsMiddleware",  # custom
     "django.contrib.sessions.middleware.SessionMiddleware",
     "django.middleware.common.CommonMiddleware",
-    "django.middleware.csrf.CsrfViewMiddleware",
+    # "django.middleware.csrf.CsrfViewMiddleware",
```
Comment (mlodic) [inline intel_owl/settings/django.py:11]: this is to have the form working correctly, am I right? We should try to enable it again, I see if I can help in any way

File: frontend/src/components/jobs/result/utils/SaveAsPlaybooksForm.jsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                />
+              </div>
+
+              <div className="p-3">
+                <Label className="required" for="forminput-name" md={12}>
+                  Playbook description
+                </Label>
+                <Input
+                  autoFocus
+                  id="forminput-description"
```
Comment (mlodic) [inline frontend/src/components/jobs/result/utils/SaveAsPlaybooksForm.jsx:78]: maybe some more space for the description?

File: api_app/playbooks_manager/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (0x0elliot) [reply api_app/playbooks_manager/models.py:19]: ahh, right. good point. totally missed that. would have caused a bug eventually.

File: api_app/playbooks_manager/serializers.py [hunk 4]
```diff
@@ -32,3 +43,142 @@ class PlaybookConfigSerializer(AbstractConfigSerializer):
...
+
+    def validate(self, attrs: dict) -> dict:
+        attrs = super().validate(attrs)
+        playbook_name = attrs["name"].replace(" ", "_").upper()
+        job_id = attrs.get("job_id")
+
+        # it might be safer for us to
+        # consider organisational permissions
+        # for accessing jobs and saving them as
+        # playbooks.
```
[the remark this answers is not in the document]
Comment (0x0elliot) [reply api_app/playbooks_manager/serializers.py:114]: for the time being, i will make it user-specific. would create another PR for adding organizational support. would have to add some small things further for that.

File: intel_owl/settings/django.py (hunk 2 above)
Comment (0x0elliot) [reply intel_owl/settings/django.py:11]: the postman had some issues with the APIs. i will enable it again. should work.

**Comment by 0x0elliot** [comment]
Screenshot 2022-11-17 at 5 51 17 AM

**Comment by mlodic** [comment]
I erronously made a commit here instead of the develop branch but noproblem, it is just a bump

**Comment by 0x0elliot** [comment]
Ready for another final review.

**Review by mlodic — COMMENTED** [review]
great! thank you very much for your effort

**Comment by mlodic** [comment]
I added some docs then I think we are good to merge

**Comment by 0x0elliot** [comment]
okay! let me know if i can help with anything else for it to merge up.

**Pull request merged**